cpp/src/io/orc/stripe_enc.cu

```diff
@@ -1260,6 +1261,39 @@ __global__ void __launch_bounds__(1024)
   }
 }
 
+struct decimal_column_element_sizes {
+  uint32_t col_idx;
+  device_span<uint32_t> sizes;
+  decimal_column_element_sizes(uint32_t col_idx, device_span<uint32_t> sizes)
+    : col_idx(col_idx), sizes(sizes)
+  {
+  }
+};
+
+template <int block_size>
+__global__ void decimal_sizes_to_offsets_kernel(device_2dspan<rowgroup_rows const> rg_bounds,
+                                                device_span<decimal_column_element_sizes> sizes)
+{
+  using block_scan = cub::BlockScan<uint32_t, block_size>;
+  __shared__ typename block_scan::TempStorage scan_storage;
+  int const t = threadIdx.x;
+
+  auto const& col_elem_sizes = sizes[blockIdx.x];
+  auto const& row_group      = rg_bounds[blockIdx.y][col_elem_sizes.col_idx];
+  auto const elem_sizes      = col_elem_sizes.sizes.data() + row_group.begin;
+
+  uint32_t initial_value = 0;
+  // do a series of block sums, storing results in the array as we go
+  for (int pos = 0; pos < row_group.size(); pos += block_size) {
+    auto const tidx    = pos + t;
+    auto tval          = tidx < row_group.size() ? elem_sizes[tidx] : 0u;
+    uint32_t block_sum = 0;
+    block_scan(scan_storage).InclusiveSum(tval, tval, block_sum);
+    if (tidx < row_group.size()) { elem_sizes[tidx] = tval + initial_value; }
+    initial_value += block_sum;
```

Comment:
Interesting, I've seen this mostly with the last thread writing this value into shared memory. I didn't know cub offered the block sum to each thread. I wonder if that impacts performance at all.

Reply:
I'm not sure. I stumbled around a bit while implementing this, it's possible there is a better way. We do have similar uses in places like `page_bounds` and `gpuUpdatePageSizes`.

Reply:
Oh, I don't think there is anything wrong with this code at all. I'm just wondering out loud.

They both have to synchronize at some point. One does it before exiting the cub call and one does it at the bottom of the loop. I would think 6 of one type of situation.